### ISSUE TITLE
agentic: wire GRPO end to end on the toy game and instrument group degeneracy

### ISSUE THREAD

**Issue by asuzukosi** [asuzukosi/ai-experiments#21]
**This closes the thin end-to-end path** — environment, rollout, reward and training all working together on the toy task before ALFWorld or an A100 is involved.

Track group degeneracy from step 0. A sparse terminal reward on a 1.5B agent that fails nearly every task means most groups are all-zero-reward: **training will look like it is running and almost nothing will be learning.** Reporting the zero-advantage fraction next to the success rate is what makes a stalled run diagnosable rather than merely disappointing.

### Sketch

```python
# src/agentic/grpo.py
episodes = [rollout(env, policy, seed=s) for s in seeds for _ in range(8)]   # G=8 per seed
rewards  = torch.tensor([e.reward for e in episodes]).view(len(seeds), 8)
adv      = rewards - rewards.mean(dim=1, keepdim=True)                       # group-relative

wandb.log({
    "zero_advantage_frac": (adv.abs().max(dim=1).values < 1e-8).float().mean().item(),
    "entropy":             policy_entropy,
    "adv_abs_p95":         adv.abs().quantile(0.95).item(),
    "clip_frac":           clip_fraction,          # drifting from ~0 == moving too fast
    "success_rate":        rewards.gt(0).float().mean().item(),
    "episode_len_mean":    mean(e.steps for e in episodes),
    "invalid_action_rate": mean(e.invalid for e in episodes),
    "truncated_frac":      mean(e.truncated for e in episodes),
})
# advantage is broadcast across the episode's ACTION tokens only (see rollout masking)
```

### TODO
- [ ] G=8 trajectories per task seed at temperature 1.0
- [ ] Trajectory reward broadcast across the episode's action tokens; group-relative advantage
- [ ] KL to a frozen reference copy of the policy; lr 1e-6; ~300-500 steps
- [ ] `scripts/train_grpo.py --reward {sparse,shaped}`
- [ ] Log zero-advantage fraction, entropy, advantage magnitude, clip fraction from step 0
- [ ] Log success rate, mean episode length, invalid-action rate and truncated fraction
- [ ] Set the rollout token budget explicitly and record it with every result
- [ ] **Done when:** 20 steps run end to end on the toy env and the toy task trains toward ~100%

### References
- TRL GRPOTrainer: https://huggingface.co/docs/trl/main/en/grpo_trainer
- verifiers multi-turn GRPO: https://github.com/willccbb/verifiers
- GRPO / group-relative advantage, DeepSeek-R1: https://arxiv.org/abs/2501.12948

### Compute

1x RTX 4090 24GB for the toy run. ~2 hours.

---

### Dependencies

- **Blocked by:** #19, #20
- **Blocks:** #36

Self-contained within this project — no cross-project dependencies.
